### PR TITLE
FPGA: Add hardware_reuse code sample

### PR DESCRIPTION
**Pull request by juny-lim**
# Adding a New Sample(s)

## Description
This sample is a tutorial that demonstrates how to use looping and task sequence to achieve FPGA hardware reuse.

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): @yuguen-intel
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

Regtest passed on
Linux https://spetc.intel.com/testanalysis?trview=0&testRunIds=10094871&tapgsize=20
Windows https://spetc.intel.com/testanalysis?trview=0&testRunIds=10094870&tapgsize=20

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/README.md [hunk 1]
```diff
@@ -85,7 +85,7 @@ flowchart LR
...
 | [hostpipes (experimental)](Tutorials/Features/experimental/hostpipes)                                                         | [Tutorials/Features](Tutorials/Features)                                                 | How to use host pipes to send and receive data between a host and the FPGA 
-| [invocation_interfaces](Tutorials/Features/hls_flow_interfaces/invocation_interfaces/)                                    | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces/)                                 | Demonstrates how to specify the kernel invocation interface and kernel argument interface for an FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler. 
+| [invocation_interfaces](Tutorials/Features/hls_flow_interfaces/invocation_interfaces/)                                        | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces/)                                 | Demonstrates how to specify the kernel invocation interface and kernel argument interface for an FPGA IP produced with the Intel® oneAPI DPC++/C++ Compiler.
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/README.md:88]: What is being changed here?

File: DirectProgramming/C++SYCL_FPGA/README.md [hunk 2]
```diff
@@ -94,14 +94,13 @@ flowchart LR
...
 | [memory_attributes](Tutorials/Features/memory_attributes)                                                                     | [Tutorials/Features](Tutorials/Features)                                                 | The basic concepts of on-chip memory attributes <br> How to apply memory attributes in your program <br> How to confirm that the memory attributes were respected by the compiler <br> A case study of the type of performance/area trade-offs enabled by memory attributes
-| [mmhost](Tutorials/Features/hls_flow_interfaces/mmhost)                                                                   | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces) | Basics of declaring Avalon memory-mapped host data interfaces for IP components
+| [mmhost](Tutorials/Features/hls_flow_interfaces/mmhost)                                                                       | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces) | Basics of declaring Avalon memory-mapped host data interfaces for IP components
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/README.md:97]: and for all of the hls_flow_interfaces samples?

File: DirectProgramming/C++SYCL_FPGA/README.md [hunk 3]
```diff
@@ -94,14 +94,13 @@ flowchart LR
...
-| [parallel_loops](Tutorials/Features/task_sequence/parallel_loops)                                          | [Tutorials/Features](Tutorials/Features) | How to use task sequences to describe multiple parallel loops in a single kernel
-
-
+| [invocation_interfaces](Tutorials/Features/hls_flow_interfaces/invocation_interfaces)                                         | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces) | How to specify the kernel invocation interface and kernel argument interfaces
+| [streaming_data_interfaces](Tutorials/Features/hls_flow_interfaces/streaming_data_interfaces)                                 | [Tutorials/Features/hls_flow_interfaces](Tutorials/Features/hls_flow_interfaces) | How to use pipes to implement streaming data interfaces on an IP component
+| [parallel_loops](Tutorials/Features/task_sequence/parallel_loops)                                                             | [Tutorials/Features](Tutorials/Features) | How to use task sequences to describe multiple parallel loops in a single kernel
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/README.md:102]: Actually, these are supposedly order alphabetically, I see that some others are not correctly placed

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+# Use cmake -DFPGA_DEVICE=<board-support-package>:<board-variant> to choose a
+# different device.
+# Note that depending on your installation, you may need to specify the full 
+# path to the board support package (BSP), this usually is in your install 
+# folder.
+#
+# You can also specify a device family (E.g. "Arria10" or "Stratix10") or a
+# specific part number (E.g. "10AS066N3F40E2SG") to generate a standalone IP.
+if(NOT DEFINED FPGA_DEVICE)
+    set(FPGA_DEVICE "Agilex7")
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt:45]:
```suggestion
    set(FPGA_DEVICE "Agilex7")
    message(STATUS "FPGA_DEVICE was not specified.\
                    \nConfiguring the design to the default FPGA family target: ${FPGA_DEVICE}\
                    \nPlease refer to the README for information on target selection.")
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+# compilation.
+set(USER_INCLUDE_PATHS ../../../../include;${USER_INCLUDE_PATHS})
+
+###############################################################################
+### no changes after here
+###############################################################################
+
+# Print the device being used for the compiles
+message(STATUS "Configuring the design to run on FPGA board ${FPGA_DEVICE}")
+
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt:65]:
```suggestion

```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    if(WIN32)
+        set(DEBUG_FLAGS /DEBUG /Od)
+    else()
+        set(DEBUG_FLAGS -g -O0)
+    endif()
+else()
+    set(DEBUG_FLAGS "")
+endif()
+
+set(COMMON_COMPILE_FLAGS -fintelfpga -Wall ${WIN_FLAG} ${DEBUG_FLAGS} ${QACTYPES} ${USER_FLAGS})
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt:112]:
```suggestion
set(COMMON_COMPILE_FLAGS -fintelfpga -Wall ${WIN_FLAG} ${QACTYPES} ${USER_FLAGS})
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+# 2. The "link" stage invokes the compiler's FPGA backend before linking. For
+#    this reason, FPGA backend flags must be passed as link flags in CMake.
+set(EMULATOR_COMPILE_FLAGS -DFPGA_EMULATOR)
+set(EMULATOR_LINK_FLAGS )
+set(REPORT_COMPILE_FLAGS -DFPGA_HARDWARE)
+set(REPORT_LINK_FLAGS -Xshardware -Xstarget=${FPGA_DEVICE} ${USER_FPGA_FLAGS} -fsycl-link=early)
+set(SIMULATOR_COMPILE_FLAGS -Xssimulation -DFPGA_SIMULATOR)
+set(SIMULATOR_LINK_FLAGS -Xssimulation -Xsghdl -Xstarget=${FPGA_DEVICE} ${USER_FPGA_FLAGS} -reuse-exe=${CMAKE_BINARY_DIR}/${SIMULATOR_OUTPUT_NAME})
+set(FPGA_COMPILE_FLAGS -DFPGA_HARDWARE)
+set(FPGA_LINK_FLAGS -Xshardware -Xstarget=${FPGA_DEVICE} ${USER_FPGA_FLAGS} -reuse-exe=${CMAKE_BINARY_DIR}/${FPGA_OUTPUT_NAME})
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/task_sequence/hardware_reuse/CMakeLists.txt:127]:
```suggestion
set(FPGA_LINK_FLAGS -Xshardware -Xstarget=${FPGA_DEVICE} ${USER_FPGA_FLAGS} -reuse-exe=${CMAKE_BINARY_DIR}/${FPGA_OUTPUT_NAME}${EXT})
```

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/README.md (hunk 1 above)
Comment (juny-lim) [reply DirectProgramming/C++SYCL_FPGA/README.md:88]: adding spaces to align the table divider line "|"

File: DirectProgramming/C++SYCL_FPGA/README.md (hunk 2 above)
Comment (juny-lim) [reply DirectProgramming/C++SYCL_FPGA/README.md:97]: adding spaces to align the table divider line "|"

File: DirectProgramming/C++SYCL_FPGA/README.md (hunk 3 above)
Comment (juny-lim) [reply DirectProgramming/C++SYCL_FPGA/README.md:102]: i did not put it in alphabetical order, I tends to add new tutorial to the end of the list. let me correct that

**Review by juny-lim — COMMENTED** [review]
I will rearrange the table in alphabetical order, and align the table divider.

**Review by tomlenth — APPROVED** [review]
Well written and documented.

**Review by yuguen-intel — APPROVED** [review]

**Pull request merged**